python/ray/tune/schedulers/pbt.py

```diff
@@ -320,34 +337,85 @@ def on_trial_result(self, trial_runner, trial, result):
         time = result[self._time_attr]
         state = self._trial_state[trial]
 
+        # Continue training if perturbation interval has not been reached yet.
         if time - state.last_perturbation_time < self._perturbation_interval:
             return TrialScheduler.CONTINUE  # avoid checkpoint overhead
 
+        # This trial has reached its perturbation interval
         score = self._metric_op * result[self._metric]
         state.last_score = score
-        state.last_perturbation_time = time
-        lower_quantile, upper_quantile = self._quantiles()
+        state.last_train_time = time
+        state.last_result = result
+
+        if not self._synch:
+            state.last_perturbation_time = time
+            lower_quantile, upper_quantile = self._quantiles()
+            self._perturb_trial(trial, trial_runner, upper_quantile,
+                                lower_quantile)
+            for trial in trial_runner.get_trials():
+                if trial.status in [Trial.PENDING, Trial.PAUSED]:
+                    return TrialScheduler.PAUSE  # yield time to other trials
 
+            return TrialScheduler.CONTINUE
+        else:
+            # Synchronous mode.
+            if any(self._trial_state[t].last_train_time <
+                   self._next_perturbation_sync and t != trial
+                   for t in trial_runner.get_trials()):
+                logger.debug("Pausing trial {}".format(trial))
+            else:
+                # All trials are synced at the same timestep.
+                lower_quantile, upper_quantile = self._quantiles()
+                all_trials = trial_runner.get_trials()
+                not_in_quantile = []
+                for t in all_trials:
+                    if t not in lower_quantile and t not in upper_quantile:
+                        not_in_quantile.append(t)
+                # Move upper quantile trials to beginning and lower quantile
+                # to end. This ensures that checkpointing of strong trials
+                # occurs before exploiting of weaker ones.
+                all_trials = upper_quantile + not_in_quantile + lower_quantile
+                for t in all_trials:
+                    logger.debug("Perturbing Trial {}".format(t))
+                    self._trial_state[t].last_perturbation_time = time
+                    self._perturb_trial(t, trial_runner, upper_quantile,
+                                        lower_quantile)
+
+                all_train_times = [
+                    self._trial_state[trial].last_train_time
+                    for trial in trial_runner.get_trials()
+                ]
+                max_last_train_time = max(all_train_times)
+                self._next_perturbation_sync = max(
+                    self._next_perturbation_sync + self._perturbation_interval,
+                    max_last_train_time)
+            return TrialScheduler.PAUSE
```

Comment:
why do we return PAUSE here? can you make a note?

Reply:
We just leave all the trials as paused here, and `choose_trial_to_run` is responsible for picking the next trial to run.